### PR TITLE
Somil55/priority queue based dijkstra

### PR DESCRIPTION
**Pull request by somil55**
Changing from Heap to priority queue in dijkstra shortest path

### CODE REVIEW

**Comment by somil55** [comment]
Benchmarks of SimpleGraph using Priority queue based dijkstra
```
julia> @benchmark dijkstra_shortest_paths(g,2)
BenchmarkTools.Trial: 
  memory estimate:  2.52 MiB
  allocs estimate:  25
  --------------
  minimum time:     376.165 μs (0.00% GC)
  median time:      499.574 μs (0.00% GC)
  mean time:        553.203 μs (9.38% GC)
  maximum time:     1.589 ms (50.42% GC)
  --------------
  samples:          8943
  evals/sample:     1

julia> @benchmark dijkstra_shortest_paths_new(g,2)
BenchmarkTools.Trial: 
  memory estimate:  1.57 MiB
  allocs estimate:  26
  --------------
  minimum time:     231.174 μs (0.00% GC)
  median time:      288.409 μs (0.00% GC)
  mean time:        347.544 μs (12.80% GC)
  maximum time:     1.518 ms (66.92% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by somil55** [comment]
Benchmarks of SimpleWeightedGraph using Priority queue based Dijkstra
```
julia> @benchmark dijkstra_shortest_paths(swg,2)
BenchmarkTools.Trial: 
  memory estimate:  2.52 MiB
  allocs estimate:  25
  --------------
  minimum time:     390.761 μs (0.00% GC)
  median time:      535.961 μs (0.00% GC)
  mean time:        587.324 μs (9.58% GC)
  maximum time:     1.646 ms (49.95% GC)
  --------------
  samples:          8426
  evals/sample:     1

julia> @benchmark dijkstra_shortest_paths_new(swg,2)
BenchmarkTools.Trial: 
  memory estimate:  1.57 MiB
  allocs estimate:  26
  --------------
  minimum time:     251.429 μs (0.00% GC)
  median time:      314.844 μs (0.00% GC)
  mean time:        371.990 μs (12.78% GC)
  maximum time:     1.586 ms (63.84% GC)
  --------------
  samples:          10000
  evals/sample:     1

```

**Comment by somil55** [comment]
Benchmarks of SimpleGraph using heap based Dijkstra
```
julia> using LightGraphs
INFO: Recompiling stale cache file /home/somil55/.julia/lib/v0.6/LightGraphs.ji for module LightGraphs.

julia> g = loadgraph("mygraph.jgz")
{40000, 40000} undirected simple Int64 graph

julia> using BenchmarkTools

julia> @benchmark
@benchmark     @benchmarkable
julia> @benchmark dijkstra_shortest_paths(g,2)
BenchmarkTools.Trial: 
  memory estimate:  2.18 MiB
  allocs estimate:  21
  --------------
  minimum time:     198.209 μs (0.00% GC)
  median time:      303.669 μs (0.00% GC)
  mean time:        349.035 μs (12.94% GC)
  maximum time:     1.086 ms (60.71% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by sbromberger** [comment]
Thanks, @somil55 ! 

TL; DR: the new PQ-based dijkstra implementation is as fast as the current one but works correctly with weighted graphs. The "heap + boolean" one also works with weighted graphs but is significantly slower.
